File: 01-Computer-Science/code/python/seb_cs/bits.py
from __future__ import annotations


def bit_at(value: int, index: int) -> int:
    if index < 0 or index > 31:
        raise ValueError("bit index out of range")
    return (value >> index) & 1


def set_bit(value: int, index: int, bit: int) -> int:
    if index < 0 or index > 31:
        raise ValueError("bit index out of range")
    if bit not in (0, 1):
        raise ValueError("bit must be 0 or 1")
    return value | (1 << index) if bit else value & ~(1 << index)


def to_binary_string(value: int, width: int = 8) -> str:
    if width < 1 or width > 32:
        raise ValueError("width out of range")
    return format(value & ((1 << width) - 1), f"0{width}b")


def u16_to_bytes(value: int, endian: str = "le") -> bytes:
    if value < 0 or value > 0xFFFF:
        raise ValueError("u16 out of range")
    return value.to_bytes(2, "little" if endian == "le" else "big")


def bytes_to_u16(data: bytes, offset: int = 0, endian: str = "le") -> int:
    chunk = data[offset : offset + 2]
    if len(chunk) < 2:
        raise ValueError("buffer too short for u16")
    return int.from_bytes(chunk, "little" if endian == "le" else "big")


def u32_to_bytes(value: int, endian: str = "le") -> bytes:
    if value < 0 or value > 0xFFFFFFFF:
        raise ValueError("u32 out of range")
    return value.to_bytes(4, "little" if endian == "le" else "big")


def bytes_to_u32(data: bytes, offset: int = 0, endian: str = "le") -> int:
    chunk = data[offset : offset + 4]
    if len(chunk) < 4:
        raise ValueError("buffer too short for u32")
    return int.from_bytes(chunk, "little" if endian == "le" else "big")

File: 01-Computer-Science/code/typescript/src/bits.ts
/** Bit/byte helpers and endian-aware integer codecs. */

export function bitAt(value: number, index: number): 0 | 1 {
  if (index < 0 || index > 31) throw new RangeError("bit index out of range");
  return ((value >>> index) & 1) as 0 | 1;
}

export function setBit(value: number, index: number, bit: 0 | 1): number {
  if (index < 0 || index > 31) throw new RangeError("bit index out of range");
  return bit ? value | (1 << index) : value & ~(1 << index);
}

export function toBinaryString(value: number, width = 8): string {
  if (width < 1 || width > 32) throw new RangeError("width out of range");
  return (value >>> 0).toString(2).padStart(width, "0").slice(-width);
}

export type Endian = "le" | "be";

export function u16ToBytes(value: number, endian: Endian = "le"): Uint8Array {
  if (value < 0 || value > 0xffff) throw new RangeError("u16 out of range");
  const out = new Uint8Array(2);
  if (endian === "le") {
    out[0] = value & 0xff;
    out[1] = (value >>> 8) & 0xff;
  } else {
    out[0] = (value >>> 8) & 0xff;
    out[1] = value & 0xff;
  }
  return out;
}

export function bytesToU16(bytes: Uint8Array, offset = 0, endian: Endian = "le"): number {
  if (offset + 2 > bytes.length) throw new RangeError("buffer too short for u16");
  return endian === "le"
    ? bytes[offset]! | (bytes[offset + 1]! << 8)
    : (bytes[offset]! << 8) | bytes[offset + 1]!;
}

export function u32ToBytes(value: number, endian: Endian = "le"): Uint8Array {
  if (value < 0 || value > 0xffffffff) throw new RangeError("u32 out of range");
  const out = new Uint8Array(4);
  let v = value >>> 0;
  if (endian === "le") {
    out[0] = v & 0xff;
    out[1] = (v >>> 8) & 0xff;
    out[2] = (v >>> 16) & 0xff;
    out[3] = (v >>> 24) & 0xff;
  } else {
    out[0] = (v >>> 24) & 0xff;
    out[1] = (v >>> 16) & 0xff;
    out[2] = (v >>> 8) & 0xff;
    out[3] = v & 0xff;
  }
  return out;
}

export function bytesToU32(bytes: Uint8Array, offset = 0, endian: Endian = "le"): number {
  if (offset + 4 > bytes.length) throw new RangeError("buffer too short for u32");
  return endian === "le"
    ? (bytes[offset]! |
        (bytes[offset + 1]! << 8) |
        (bytes[offset + 2]! << 16) |
        (bytes[offset + 3]! << 24)) >>>
        0
    : ((bytes[offset]! << 24) |
        (bytes[offset + 1]! << 16) |
        (bytes[offset + 2]! << 8) |
        bytes[offset + 3]!) >>>
        0;
}
